File: 6. MostWater.js
//Leet Code: 11 : Container with Most Water: https://leetcode.com/problems/container-with-most-water/description/

//Approach1: O(n)
//Step1: create variables for start and end.
//Step2: calculate volume between the start and end index.
//Step3: Volume = smaller-height * abs(i-j)
//Step4: Compare volume with max volume.
//Step5: Return the max volume.

function containerWithMostWater(height){
    let left = 0;
    let right = height.length -1;
    let volume = 0;
    let maxVolume = 0;
    while (left < right){
        volume = Math.min(height[left], height[right])*Math.abs(left-right);
        if(volume > maxVolume){
            maxVolume = volume;
        }

        if(height[left] > height[right]){
            right --;
        }
        else{
            left ++;
        }
    }
    return maxVolume;
}

console.log(containerWithMostWater([1,1]));

File: 5. Two-Pointer/6. MostWater.py
#Leet Code: 11 : 
# Container with Most Water: https://leetcode.com/problems/container-with-most-water/description/

# Approach: Two Pointer: T: O(n) | O(1)
# Step 1: Use left and right pointer 
# Step 2: Compute Volume and Max Volume
# Step 3: Decrease left and right based on height

class Solution:
    def maxArea(height: list[int]) -> int:
        n = len(height)
        left = 0
        right = n - 1
        volume = 0
        max_volume = 0

        while(left < right):
            d = right - left
            h = min(height[left], height[right])
            volume = d * h

            if(volume > max_volume):
                max_volume = volume
            
            if(height[left] == height[right]):
                left += 1
                right -= 1
            elif(height[left] > height[right]):
                right -= 1
            else: 
                left += 1
        print(max_volume)
        return max_volume


result = Solution()

result1 = Solution.maxArea([1,8,6,2,5,4,8,3,7])


